src/policy/rbf.cpp

```diff
@@ -181,3 +181,54 @@ std::optional<std::string> PaysForRBF(CAmount original_fees,
     }
     return std::nullopt;
 }
+
+std::optional<std::string> CheckConflictTopology(const CTxMemPool::setEntries& direct_conflicts)
+{
+    for (const auto& direct_conflict : direct_conflicts) {
+        const auto desc_count = direct_conflict->GetCountWithDescendants() - 1;
+        if (desc_count == 0) {
+            // It should be by itself, or be a child in a parent-child cluster
+            if (direct_conflict->GetCountWithAncestors() > 2) {
+                return strprintf("Child transaction has too many ancestors");
+            }
+            if (direct_conflict->GetCountWithAncestors() == 2) {
+                const auto& parent = direct_conflict->GetMemPoolParentsConst().begin();
+                if (parent->get().GetCountWithDescendants() > 2) {
+                    return strprintf("Parent in package rbf should not have multiple descendants");
+                }
+            }
+        } else if (desc_count == 1) {
+            // It should be the parent in a parent-child cluster
+            const auto& parent = direct_conflict;
+            if (parent->GetCountWithAncestors() > 1) {
+                return strprintf("Parent in package rbf should not have any ancestors");
+            }
+            const auto& child = parent->GetMemPoolChildrenConst().begin();
+            if (child->get().GetCountWithAncestors() != 2) {
+                return strprintf("Child in package RBF has too many ancestors");
+            }
+        } else {
+            return strprintf("Too many descendants of direct conflict: %zu", desc_count);
+        }
+    }
+    return std::nullopt;
+}
+
+std::optional<std::string> CheckMinerScores(CAmount replacement_fees,
+                                            int64_t replacement_vsize,
+                                            const CTxMemPool::setEntries& original_transactions)
+{
+        // Note that this assumes no in-mempool ancestors
+        const CFeeRate replacement_miner_score(replacement_fees, replacement_vsize);
+
+        for (const auto& entry : original_transactions) {
+            CFeeRate original_miner_score(entry->GetModFeesWithAncestors(), entry->GetSizeWithAncestors());
+            original_miner_score = std::min(original_miner_score, CFeeRate(entry->GetModifiedFee(), entry->GetTxSize()));
+            if (replacement_miner_score < original_miner_score) {
```

Comment:
Shouldn’t this be greater or equal to _exceed_ the original?

> 6. The package feerate (total package fee / total package vsize) must exceed the min(individual feerate, ancestor feerate) of every transaction that would be evicted (direct and indirect conflicts).

Reply:
Probably yeah, tests should catch this case regardless. Will add one after deciding what way to go with heuristic vs diagram check

Reply:
replaced with diagram check; should be fine now since it must be strictly superior.